File: helpers/testing/graphs/graphSearch.js
const breadthFirstSearch = (root, id) => {
	if (!root) return undefined;

	const q = [];
	const visited = new Set();

	q.push(root);

	while (q.length) {
		const n = q.shift();

		if (n.id === id) return n;

		visited.add(n);

		for (const child of n.children) {
			if (!visited.has(child)) {
				q.push(child);
			}
		}
	}

	return undefined;
};

module.exports = breadthFirstSearch;

File: python/helpers/testing/graph_search.py
from typing import Optional

from helpers.graph_classes.graph_node import GraphNode


def breadth_first_search(
    root: Optional[GraphNode], identifier: int
) -> Optional[GraphNode]:
    if root is None:
        return None

    q = []
    visited = set()

    q.append(root)

    while len(q) > 0:
        n = q.pop(0)
        visited.add(n)

        if n.get_id() == identifier:
            return n

        for child in n.get_children():
            if child not in visited:
                q.append(child)

    return None
